### ISSUE TITLE
Invalid variable name: variable cannot start with numbers. Breaking change introduced

### ISSUE THREAD

**Issue by maxadamo** [hashicorp/terraform#21803]
### Terraform Version
0.12.2

### Terraform Configuration Files

```hcl
variable "1" {
  type = "map"
...
```

### Crash Output
```
Error: Invalid variable name
```
### Expected Behavior
Accept numbers, as it was until version 0.11.*

### Actual Behavior
```
Error: Invalid variable name

  on variables.tf line 28, in variable "1":
  28: variable "1" {

A name must start with a letter and may contain only letters, digits,
underscores, and dashes.
```

### Additional Context
I understand that it complies more with rules of several programming languages, but I used to generate these numbers from a template, and it worked until I switched to version 0.12

**Comment by invidian** [comment]
There was #19919, but I indeed cannot see that being mentioned in 0.12.0 changelog as a breaking change.

**Comment by apparentlymart** [comment]
Hi @maxadamo,

The upgrade guide [focuses on resource names in particular](https://www.terraform.io/upgrade-guides/0-12.html#pre-upgrade-checklist), but indeed in general 0.12 fixed all of the validation of declared names so that none of them are now missing the validation rule to require valid identifiers. A variable named "1" is ambiguous because a reference to it could be either an index into an object or a variable name. Terraform 0.11 was inconsistent in how different parts of it treated such a reference, but Terraform 0.12 now consistently treats digits as introducing numbers and never as the start of an identifier.

Before upgrading you will need to change your generator to only generate valid variable names.

I'm sorry that the guide itself did not cover this specific detail. Because the changes in 0.12 were quite substantial, we focused primarily on documenting situations where a configuration might be misinterpreted without an error in 0.12, or situations where Terraform 0.12 could not give a clear error to explain how to fix it. In this case the error message is clear about what change needs to happen, so we did not mention it explicitly in the upgrade guide; to have listed all of these situations explicitly would have made the upgrade guide impenetrable.
